### DISCUSSION TITLE
`#[new]` and `#[classmethod]`

### DISCUSSION THREAD

**Discussion by adriangb** [PyO3/pyo3#3075] [Ideas]
As far as I can tell, `#[new]` has to be a static method. It's a classmethod in Python, could it be made the same in PyO3? Context: I'd like to get the class name as seen by Python for the failure case of calling new.

**Comment by davidhewitt** [comment]
This seems like a reasonable thing to support, I think we could just allow `#[new]` and `#[classmethod]` attributes to be combined? Or we could have `#[new(classmethod)]`?

The implementation might need a bit of refactoring to make it work, the relevant code is here: https://github.com/PyO3/pyo3/blob/77fc6e67340e5e27aac9d860d24266be6a6e1cbc/pyo3-macros-backend/src/method.rs#L504

`subtype` is exactly the `cls` which you'd want PyO3 to forward into the function. To enable that I think we'd need to carry extra information around which would probably require a bit of reworking of the macros. I'll try to see if I can get this done for 0.19, also PRs to help in this regard much appreciated 😄

**Reply by adriangb** [reply]
Thanks David! No promises but if I find some extra time I'll take a look at implementing it. I've never touched unsafe Rust though...

**Reply by birkenfeld** [reply]
In the case of PyO3, it's mostly unsafe due to FFI interaction with the C API. So if you know how to program extensions in C, you'll be alright.

**Reply by adriangb** [reply]
> So if you know how to program extensions in C, you'll be alright.

😰
